### PR TITLE
Closes #6565: Handle deny & don't ask again in SitePermissionsFeature.

### LINKED ISSUE #6565: Handle the deny & don't ask again user use case on the site permissions feature
We are not handling this user flow as is described in the Fenix related issue https://github.com/mozilla-mobile/fenix/issues/7566#issuecomment-609426117

### PR DESCRIPTION
**Pull request by Amejia481**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [x] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt [hunk 1]
```diff
@@ -484,7 +499,9 @@ class SitePermissionsFeature(
         return false
     }
 
-    private val PermissionRequest.host get() = uri?.toUri()?.host ?: ""
+    private fun PermissionRequest.getHost(session: Session) = (uri?.toUri()?.host
+            ?: "").ifEmpty { session.url.toUri().host ?: "" }
```
Comment (pocmo) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:503]: nit: Using an empty string only to then check if it is an empty string seems unnecessary. Afaik we can chain them via the nullable check? `uri?.toUri()?.host
    ?: session.url.toUri().host ?: ""`

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt [hunk 2]
```diff
@@ -74,7 +81,8 @@ class SitePermissionsFeature(
     private val fragmentManager: FragmentManager,
     var promptsStyling: PromptsStyling? = null,
     private val dialogConfig: DialogConfig? = null,
-    override val onNeedToRequestPermissions: OnNeedToRequestPermissions
+    override val onNeedToRequestPermissions: OnNeedToRequestPermissions,
+    val onShouldShowRequestPermissionRationale: (permission: String) -> Boolean
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:85]: Agree, I thought about that too, it's bit tricky as to get access to the `shouldShowRequestPermissionRationale` function, as we need a reference of an [activity](https://developer.android.com/reference/androidx/core/app/ActivityCompat#shouldShowRequestPermissionRationale(android.app.Activity,%20java.lang.String)) or a [fragment](https://developer.android.com/reference/androidx/fragment/app/Fragment#shouldShowRequestPermissionRationale(java.lang.String)). Do you have any suggestions? :)

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt (hunk 1 above)
Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:503]: :sweat_smile: 
Thanks for catching this!

Comment (Amejia481) [reply components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:503]: The only thing is bit hard to read, for the nested checks 😅, I added it as it a bit clear, but I guess it's matter of taste

**Comment by Amejia481** [comment]
bors r=pocmo

**Pull request merged**